### PR TITLE
Conditional serializing

### LINKED ISSUE #77: Create a way to partially serialise objects
When making a POST request only a few of object's properties are needed, for example, when POSTing a `Client`, only it's `Name` and `WorkspaceId` are needed (`Id` and `At`) shouldn't be posted.

This could be achieved with attributes.

### PR DESCRIPTION
**Pull request by willsb**
Closes #77 

This adds conditional serialization for the post method. It’s also extensible for other verbs, when needed.

### CODE REVIEW

File: Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+{
+        protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
+        {
+            IList<JsonProperty> properties = base.CreateProperties(type, memberSerialization);
+            foreach (JsonProperty property in properties)
+            {
+                var attributes = property.AttributeProvider.GetAttributes(typeof(TSerializationAttribute), false);
+                if (!attributes.Any()) continue;
+
+                property.ShouldSerialize = _ => false;
```
Comment (amulware) [inline Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs]: Why not simplify to
```
if (attributes.Any())
    property.ShouldSerialize = _ => false;
```
?

File: Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Toggl.Ultrawave.Serialization
+{
+    internal sealed class AttributeAwareContractResolver<TSerializationAttribute> : BaseContractResolver
```
Comment (amulware) [inline Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs]: Do I understand correctly that this resolver will make sure that properties with `TSerializationAttribute` will not be serialized? If so, the name seems to not reflect this very well.

Either change the name, or give this class an abstract method it calls for each property with the given attribute, and let a subclass decide what to do with the found properties.

File: Toggl.Ultrawave/Serialization/SerializationReason.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Toggl.Ultrawave.Serialization
+{
+    internal enum SerializationReason
```
Comment (amulware) [inline Toggl.Ultrawave/Serialization/SerializationReason.cs]: We will probably convert/wrap this into a different thing that also encapsulates user/workspace permissions, etc. in the future, correct?

File: Toggl.Ultrawave/Serialization/JsonSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+using static Toggl.Ultrawave.Serialization.SerializationReason;
+
+namespace Toggl.Ultrawave.Serialization
+{
+    internal class JsonSerializer : IJsonSerializer
+    {
+        private readonly JsonSerializerSettings defaultSettings = new TogglSerializerSettings<BaseContractResolver>();
+		
+        private readonly JsonSerializerSettings postSettings = 
+            new TogglSerializerSettings<AttributeAwareContractResolver<IgnoreWhenPostingAttribute>>();
```
Comment (amulware) [inline Toggl.Ultrawave/Serialization/JsonSerializer.cs]: This is a rather unwieldy type name. Why does the whole thing have to be generic anyways?

File: Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs (hunk 2 above)
Comment (willsb) [inline Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs]: Good point

File: Toggl.Ultrawave/Serialization/SerializationReason.cs (hunk 3 above)
Comment (willsb) [inline Toggl.Ultrawave/Serialization/SerializationReason.cs]: Why convert? We simply need to add new cases to this enum. Maybe improve naming a bit

File: Toggl.Ultrawave/Serialization/JsonSerializer.cs (hunk 4 above)
Comment (willsb) [inline Toggl.Ultrawave/Serialization/JsonSerializer.cs]: Another alternative is to have a static method like `SerializerSettings.For<TContractResolver>`

File: Toggl.Ultrawave/Serialization/SerializationReason.cs (hunk 3 above)
Comment (amulware) [inline Toggl.Ultrawave/Serialization/SerializationReason.cs]: We will likely require multi-dimensional data in the future. But we'll see that then.

**Comment by amulware** [comment]
Please fix build.

File: Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Toggl.Ultrawave.Serialization
+{
+    internal sealed class IgnoreAttributeContractResolver<TIgnoredAttribute> : DefaultContractResolver
```
Comment (amulware) [inline Toggl.Ultrawave/Serialization/AttributeAwareContractResolver.cs]: Does not match file name.

**Pull request merged**